File: user_app/lib/features/models/product/product_model.dart
import 'package:Quickzon/features/models/models.dart';
import 'package:Quickzon/core/utils/enums/enums.dart';
import 'package:json_annotation/json_annotation.dart';

part 'product_model.g.dart';

@JsonSerializable(explicitToJson: true)
class ProductType {
  @JsonKey(name: '_id')
  final String id;

  final String name;
  final String slug;
  final String? sellerId;
  final String? brand;

  // categoryId is an object in the response
  final CategoryBasic categoryId;

  // categoryPath is array of CategoryBasic objects
  @JsonKey(name: 'categoryPath', defaultValue: [])
  final List<CategoryBasic> categoryHierarchy;

  @JsonKey(defaultValue: [])
  final List<String> tags;

  @JsonKey(defaultValue: [])
  final List<String> images;

  final String? mainImage;
  final String? description;
  final String? shortDescription;
  final String? metaTitle;
  final String? metaDescription;
  final String? metaKeywords;
  final ProductEatableType productType;
  final String? manufacturer;
  final String? madeIn;
  final String? fssaiNumber;
  final String? barcode;
  final int maxQtyPerUser;
  final bool isReturnable;
  final bool isCOD;
  final bool isCancelable;
  final ProductStatus status;

  // variants maps to variants
  @JsonKey(name: 'variants', defaultValue: [])
  final List<ProductVariantType> variants;

  @JsonKey(name: 'ratings', fromJson: _ratingsAvgFromJson)
  final double ratingsAvg;

  @JsonKey(defaultValue: 0)
  final int popularity;

  @JsonKey(defaultValue: [])
  final List<String> searchKeywords;

  final TaxRateType? taxRate;
  final bool isApproved;
  final bool isActive;
  final bool isDeleted;
  final String? deletedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  @JsonKey(name: '__v')
  final int? v;

  // Computed property for backward compatibility
  CategoryBasic get category => categoryId;

  ProductType({
    required this.id,
    required this.name,
    required this.slug,
    this.sellerId,
    this.brand,
    required this.categoryId,
    required this.categoryHierarchy,
    required this.tags,
    required this.images,
    this.mainImage,
    this.description,
    this.shortDescription,
    this.metaTitle,
    this.metaDescription,
    this.metaKeywords,
    required this.productType,
    this.manufacturer,
    this.madeIn,
    this.fssaiNumber,
    this.barcode,
    required this.maxQtyPerUser,
    required this.isReturnable,
    required this.isCOD,
    required this.isCancelable,
    required this.status,
    required this.variants,
    required this.ratingsAvg,
    required this.popularity,
    required this.searchKeywords,
    this.taxRate,
    required this.isApproved,
    required this.isActive,
    required this.isDeleted,
    this.deletedAt,
    required this.createdAt,
    required this.updatedAt,
    this.v,
  });

  // Helper to parse ratings.avg with null safety
  static double _ratingsAvgFromJson(dynamic json) {
    if (json == null) return 0.0;
    if (json is Map<String, dynamic>) {
      final avg = json['avg'];
      if (avg == null) return 0.0;
      if (avg is num) return avg.toDouble();
    }
    return 0.0;
  }

  // Helper to parse ratings.count with null safety
  static int _ratingsCountFromJson(dynamic json) {
    if (json == null) return 0;
    if (json is Map<String, dynamic>) {
      final count = json['count'];
      if (count == null) return 0;
      if (count is num) return count.toInt();
    }
    return 0;
  }

  factory ProductType.fromJson(Map<String, dynamic> json) =>
      _$ProductTypeFromJson(json);

  Map<String, dynamic> toJson() => _$ProductTypeToJson(this);
}

File: server/src/models/product/product.model.ts
// product.model.ts
import mongoose from "mongoose";
import { TaxRateType, IProductDocument, IVariant, ProductEatableType, ProductModelType, ProductStatus, VariantInventoryType, VariantQuantityType, VariantStatus } from "../../lib/types/index";
import { createProductPreSave, deleteProductsByCategory, rebuildProductCategoryPaths, restoreProductsByCategory, } from "./product.utils";

const VariantSchema = new mongoose.Schema<IVariant>(
    {
        title: { type: String },
        sku: { type: String, required: true, index: true, unique: true, sparse: true },
variantType: {
  type: String,
},
        measurement: { type: Number },
        measurementUnit: { type: String },
        price: { type: Number, required: true },
        mrp: { type: Number },
        discountPercent: { type: Number, default: 0 },
        discountedPrice: { type: Number },
        stock: { type: Number, default: 0 },
        inventoryType: { type: String, enum: Object.values(VariantInventoryType), default: VariantInventoryType.LIMITED },
        status: { type: String, enum: Object.values(VariantStatus), default: VariantStatus.AVAILABLE },
        images: [{ type: String }],
    },
    { _id: true }
);

const ProductSchema = new mongoose.Schema<IProductDocument, ProductModelType>(
    {
        name: { type: String, required: true },
        slug: { type: String, required: true, unique: true },
        brandId: { type: mongoose.Schema.Types.ObjectId, ref: "Brand", default: null, },
        categoryId: { type: mongoose.Schema.Types.ObjectId, ref: "Category", required: true },
        categoryPath: [{ type: mongoose.Schema.Types.ObjectId, ref: "Category" }],
        description: { type: String },
        shortDescription: { type: String },
        metaTitle: { type: String },
        metaDescription: { type: String },
        metaKeywords: { type: String },
        tags: [{ type: String, index: true }],
        mainImage: { type: String },
        images: [{ type: String }],
        productType: { type: String, enum: Object.values(ProductEatableType), default: ProductEatableType.NONE },
        isReturnable: { type: Boolean, default: true },
        isCOD: { type: Boolean, default: true },
        isCancelable: { type: Boolean, default: true },
        maxQtyPerUser: { type: Number, default: 10 },
        taxRate: { type: String, enum: Object.values(TaxRateType), default: TaxRateType.GST_5 },
        variants: [VariantSchema],
        status: { type: String, enum: Object.values(ProductStatus), default: ProductStatus.APPROVED },
        isActive: { type: Boolean, default: false },
        isDeleted: { type: Boolean, default: false },
        deletedAt: { type: Date, default: null },

        manufacturer: { type: String },
        madeIn: { type: String },
        fssaiNumber: { type: String },
        barcode: { type: String, index: true },
        ratings: {
            avg: { type: Number, default: 0 },
            count: { type: Number, default: 0 },
        },
        popularity: { type: Number, default: 0 },
        featured: { type: Boolean, default: false },
        searchKeywords: [{ type: String }],
        sellerId: { type: mongoose.Schema.Types.ObjectId, ref: "Seller", default: null },
        isApproved: { type: Boolean, default: true },
    },
    { timestamps: true, toJSON: { virtuals: true }, toObject: { virtuals: true } }
);

// Indexes
ProductSchema.index(
    { name: "text", description: "text", tags: "text", brand: "text", metaTitle: "text", metaKeywords: "text" },
    { weights: { name: 10, brand: 5, tags: 3, description: 2 } }
);
ProductSchema.index({ categoryId: 1, isActive: 1, isDeleted: 1 });
ProductSchema.index({ sellerId: 1, isActive: 1 });
ProductSchema.index({ "variants.price": 1 });
ProductSchema.index({ "variants.stock": 1 });
ProductSchema.index({ createdAt: -1 });
ProductSchema.index({ popularity: -1 });
ProductSchema.index({ searchKeywords: 1 });

// Pre-save hooks
ProductSchema.pre("save", createProductPreSave);
ProductSchema.virtual('brand', { ref: 'Brand', localField: 'brandId', foreignField: '_id', justOne: true });
ProductSchema.static("deleteProductsByCategory", deleteProductsByCategory)
ProductSchema.static("restoreProductsByCategory", restoreProductsByCategory)
ProductSchema.static("rebuildProductCategoryPaths", rebuildProductCategoryPaths)

export const Product = mongoose.model<IProductDocument, ProductModelType>("Product", ProductSchema);
